src/mono/mono/metadata/w32mutex-unix.c

```diff
@@ -253,180 +250,41 @@ mono_w32mutex_init (void)
-	gpointer handle = mono_w32handle_namespace_search_handle (MONO_W32TYPE_NAMEDMUTEX, utf8_name);
-
-	mono_w32handle_namespace_unlock ();
-
-	if (handle == INVALID_HANDLE_VALUE) {
-		/* The name has already been used for a different object. */
-		*win32error = ERROR_INVALID_HANDLE;
-		return handle;
-	} else if (!handle) {
-		/* This name doesn't exist */
-		*win32error = ERROR_FILE_NOT_FOUND;
-		return handle;
-	}
-
-	mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_IO_LAYER_MUTEX, "%s: returning named mutex handle %p",
-		__func__, handle);
-
-	return handle;
-}
+#ifndef HOST_WIN32
+static GENERATE_GET_CLASS_WITH_CACHE (wait_subsystem, "System.Threading", "WaitSubsystem");
 
 void
 mono_w32mutex_abandon (MonoInternalThread *internal)
 {
-	// Thread and InternalThread are pinned/mature.
-	// Take advantage of that and do not use handles here.
-	g_assert (mono_thread_internal_is_current (internal));
-
-	if (!internal->owned_mutexes)
-		return;
-
-	while (internal->owned_mutexes->len) {
-		MonoW32Handle *handle_data;
-		MonoW32HandleMutex *mutex_handle;
-		MonoNativeThreadId tid;
-		gpointer handle;
+	ERROR_DECL (error);
 
-		handle = g_ptr_array_index (internal->owned_mutexes, 0);
+	MONO_STATIC_POINTER_INIT (MonoMethod, thread_exiting)
 
-		if (!mono_w32handle_lookup_and_ref (handle, &handle_data))
-			g_error ("%s: unkown handle %p", __func__, handle);
+		MonoClass *wait_subsystem_class = mono_class_get_wait_subsystem_class ();
+		g_assert (wait_subsystem_class);
+		thread_exiting = mono_class_get_method_from_name_checked (wait_subsystem_class, "OnThreadExiting", -1, 0, error);
+		mono_error_assert_ok (error);
 
-		if (handle_data->type != MONO_W32TYPE_MUTEX && handle_data->type != MONO_W32TYPE_NAMEDMUTEX)
-			g_error ("%s: unkown mutex handle %p", __func__, handle);
+	MONO_STATIC_POINTER_INIT_END (MonoMethod, thread_exiting)
 
-		mutex_handle = (MonoW32HandleMutex*) handle_data->specific;
+	g_assert (thread_exiting);
 
-		mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_IO_LAYER_MUTEX, "%s: abandoning %s handle %p",
-			__func__, mono_w32handle_get_typename (handle_data->type), handle);
-
-		tid = MONO_UINT_TO_NATIVE_THREAD_ID (internal->tid);
-
-		if (!pthread_equal (mutex_handle->tid, tid))
-			g_error ("%s: trying to release mutex %p acquired by thread %p from thread %p",
-				__func__, handle, (gpointer) mutex_handle->tid, (gpointer) tid);
-
-		mono_w32handle_lock (handle_data);
-
-		mutex_handle->recursion = 0;
-		mutex_handle->tid = 0;
-		mutex_handle->abandoned = TRUE;
+	if (mono_runtime_get_no_exec ())
+		return;
 
-		mono_w32handle_set_signal_state (handle_data, TRUE, FALSE);
+	HANDLE_FUNCTION_ENTER ();
 
-		thread_disown_mutex (internal, handle);
+	ERROR_DECL (error);
 
-		mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_IO_LAYER_MUTEX, "%s: abandoned %s handle %p",
-			__func__, mono_w32handle_get_typename (handle_data->type), handle);
+	gpointer args [1];
+	args [0] = internal;
+	mono_runtime_try_invoke_handle (thread_exiting, NULL_HANDLE, args, error);
```

Comment:
`mono_w32mutex_abandon` is called from `mono_thread_detach_internal` which is called from, for example, the TLS key dtor for the MonoThreadInfo (via the `thread_detach` callback). If that happens, it's might be too late to run managed code.

Reply:
On corert, the runtime calls the OnThreadExit () callback in Thread.CoreRT.Unix.cs when the thread exists. This is called on the thread itself just before it exits.

Reply:
yea for managed threads I agree it makes sense to just call it from managed.

but for a foreign thread that just called some reverse pinvoke and doesn't know anything about Mono, if it exits while holding a mutex, our only chance to cleanup is through the TLS destructor.  I'd be curious to know what CoreRT does in that case.

Reply:
CoreRT just calls into managed code in this case. We can reason about the bit of managed code that runs in this case. We know that it has been fully AOT and there will be no JITing or other heavy weight operation.

Reply:
What we do in a lot of cases in mono is do the operation from the finalizer thread, i.e. see mono_gc_finalize_notify ().